libmy/my_config.c

```diff
@@ -0,0 +1,378 @@
+		if (strncmp(config->root->name, name, name_len) == 0)
+			return config->root;
+	}
+	section = ISC_LIST_HEAD(config->sections);
+	while(section != NULL) {
+		name_len = (section->name_len < name_len ? section->name_len : name_len);
+
+		if (strncmp(section->name, name, name_len) == 0)
+			break;
+		section = ISC_LIST_NEXT(section, link);
+	}
+
+	return section;
+}
+
+static void
+_my_config_destroy_item(struct my_config_item *item) {
+	my_free(item->item->key);
+	my_free(item->item->value);
+	my_free(item->item);
+}
+
+static void
+_my_config_destroy_items(struct _my_config *section) {
+	struct my_config_item *items;
+	items = ISC_LIST_HEAD(section->items);
+
+	while(items != NULL) {
+		struct my_config_item *next;
+		next = ISC_LIST_NEXT(items, link);
+		_my_config_destroy_item(items);
+		ISC_LIST_UNLINK(section->items, items, link);
+		my_free(items);
+		items = next;
+	}
+}
+
+static void
+_my_config_destroy_sections(struct my_config *config) {
+	struct _my_config *sections;
+
+	if (config->root != NULL) {
+		_my_config_destroy_items(config->root);
+		my_free(config->root->name);
+		my_free(config->root);
+	}
+
+	sections = ISC_LIST_HEAD(config->sections);
+
+	while(sections != NULL) {
+		struct _my_config *next;
+		next = ISC_LIST_NEXT(sections, link);
+		my_free(sections->name);
+		_my_config_destroy_items(sections);
+		ISC_LIST_UNLINK(config->sections, sections, link);
+		my_free(sections);
+		sections = next;
+	}
+}
+
+/*
+ * PUBLIC FUNCTIONS
+ */
+
+struct my_config *
+my_config_init(void) {
+	struct my_config *result;
+	result = my_calloc(1, sizeof(struct my_config));
+
+	ISC_LIST_INIT(result->sections);
+
+	return result;
+}
+
+bool
+my_config_fill(struct my_config *config, const char *data) {
+	const char *divider = NULL;
+	struct _my_config *section = NULL;
+	size_t data_len;
+
+	if (config == NULL || data == NULL || *data == '\0')
+		return false;
+
+	section = _my_config_find_section(config, MY_CONFIG_DEFAULT_SECTION, sizeof(MY_CONFIG_DEFAULT_SECTION));
+	if (section == NULL)
+		section = _my_config_add_section(config, true, MY_CONFIG_DEFAULT_SECTION, sizeof(MY_CONFIG_DEFAULT_SECTION));
+
+	for(;;) {
+		divider = strchr(data, _COLON_DIV);
+		data_len = (divider != NULL ? (size_t) (divider - data) : strlen(data));
+		if (data_len == 0)
+			break;
+		if (!_my_config_add_item(section, data, data_len))
+			return false;
+		if (divider == NULL)
+			break;
+		data += data_len + 1;
+	}
+
+	return true;
+}
+
+bool
+my_config_load(struct my_config *config, const char *filename) {
+	bool result = false;
+	char buffer[1024];
+	FILE *f;
+	struct _my_config *section = NULL;
+
+	if (config == NULL || filename == NULL)
+		return false;
+
+	f = fopen(filename, "r");
+	if (f == NULL)
+		return false;
+
+	while(fgets(buffer, sizeof(buffer), f)) {
+		size_t line_len;
+		char *ptr = buffer;
+		char *eol;
+
+		while(*ptr == ' ')
+			++ptr;
+
+		if (*ptr == _COMMENT || *ptr == '\0' || *ptr == '\n')
```

Comment:
We may not want any comments other than those that start a line... otherwise we might need more complex parsing to handle string values with a '#' inside - which is not so unlikely